device/common/include/traccc/clusterization/device/impl/ccl_kernel.ipp

```diff
@@ -244,73 +252,23 @@ TRACCC_DEVICE inline void ccl_kernel(
 
     barrier.blockBarrier();
 
-    /*
-     * Count the number of clusters by checking how many cells have
-     * themself assigned as a parent.
-     */
-    for (index_t tst = 0, cid; (cid = tst * blckDim + threadId) < size; ++tst) {
-
-        if (f[cid] == cid) {
-            // Increment the summary values in the header object.
-            vecmem::device_atomic_ref<unsigned int,
-                                      vecmem::device_address_space::local>
-                atom(outi);
-            atom.fetch_add(1);
-        }
-    }
-
-    barrier.blockBarrier();
-
-    /*
-     * Add the number of clusters of each thread block to the total
-     * number of clusters. At the same time, a cluster id is retrieved
-     * for the next data processing step.
-     * Note that this might be not the same cluster as has been treated
-     * previously. However, since each thread block spawns a the maximum
-     * amount of threads per block, this has no sever implications.
-     */
-    if (threadId == 0) {
-        vecmem::device_atomic_ref<unsigned int,
-                                  vecmem::device_address_space::global>
-            atom(measurement_count);
-        outi = atom.fetch_add(outi);
-    }
-
-    barrier.blockBarrier();
-
-    /*
-     * Get the position to fill the measurements found in this thread group.
-     */
-    const unsigned int groupPos = outi;
```

Comment:
Yeah I'm confused, where has this code gone?

Reply:
This code is trying to be super fancy with how it
  - counts the total number of clusters;
  - figures out where to put the created measurement in the output container.

I decided that since resizable vecmem buffers can do all of this out of the box anyway, I'd not put the individual measurements into specific places in the output container. And would rather push them back into the output container in a random order. The final part of the code now does that.

It definitely works correctly, as all the tests are still happy with the output of the algorithm.

But indeed, if the ordering of the measurements was important, then this PR's code is not doing the same thing as the current `main` branch does. :thinking: What exact ordering do we require from the reconstructed measurements? Is it required that ones coming from the same module would be consecutive in the container? Because I'm not sure why that would be happening even with the current code? (Since the thread-blocks can operate across modules as-is.)

And yeah, removing all this code is what made the kernel a little faster. That it now only does the

```
    for (details::index_t tst = 0, cid; (cid = tst * blckDim + threadId) < size;
         ++tst) {
```

loop once, instead of twice. (Exchanging one extra loop, atomic operations on shared variables, and a few atomic operations on a global variable, for more atomic operations on the global variable.)

Reply:
>  Is it required that ones coming from the same module would be consecutive in the container? 

Right. This is what is required from the track finding. The order iteself may not be very important.

Reply:
> Since the thread-blocks can operate across modules as-is.

This is not point of sorting. Unless the measurements in the same module are consecutive, it will be very difficult to iterate over the measurements in the same module during the CKF.

Reply:
Did some checking with the compute sanitizer, seems like the following is popping up frequently:

```
========= Invalid __global__ read of size 4 bytes
=========     at traccc::device::reduce_problem_cell(const vecmem::device_vector<const traccc::cell> &, unsigned short, unsigned int, unsigned int, unsigned char &, unsigned short *)+0x510 in /mnt/ssd1/sswatman/traccc/device/common/include/traccc/clusterization/device/impl/reduce_problem_cell.ipp:25
=========     by thread (28,0,0) in block (0,0,0)
=========     Address 0x3ca7a10 is out of bounds
=========     and is 139,622,790,563,312 bytes before the nearest allocation at 0x7efc7aa00000 of size 4 bytes
=========     Device Frame:void traccc::device::ccl_kernel<traccc::cuda::barrier>(unsigned short, unsigned short, unsigned int, vecmem::data::vector_view<const traccc::cell>, vecmem::data::vector_view<const traccc::cell_module>, unsigned short, unsigned short, unsigned int &, unsigned int &, unsigned int &, unsigned short *, unsigned short *, T1 &, vecmem::data::vector_view<traccc::measurement>, vecmem::data::vector_view<unsigned int>)+0x2fc0 in /mnt/ssd1/sswatman/traccc/device/common/include/traccc/clusterization/device/impl/ccl_kernel.ipp:226
=========     Device Frame:traccc::cuda::kernels::ccl_kernel(vecmem::data::vector_view<const traccc::cell>, vecmem::data::vector_view<const traccc::cell_module>, unsigned short, unsigned short, vecmem::data::vector_view<traccc::measurement>, vecmem::data::vector_view<unsigned int>)+0x15b0 in /mnt/ssd1/sswatman/traccc/device/cuda/src/clusterization/clusterization_algorithm.cu:40
```

Reply:
There were two issues with the unit tests:
   - I didn't add correct host->device->host copies to the CUDA test when switching it to use `traccc::cuda::clusterization_algorithm`;
   - I switched the clusterization code to using `traccc::details::is_far_enough(...)` in an effort to increase the code sharing between the CPU and GPU codebases. But as it turns out, the existing function (which works fine for the host algorithm) was not working well for the FastSV algorithm. So now I switched it to a setup that does.

Reply:
Great, looks like everything is in order functionality-wise now; thanks!

However, it also seems to bring some scary performance degradation. Before this PR (commit `3b587755`):

```
[2379250] traccc_throughput_mt_cuda@127.0.0.1
  traccc::cuda::kernels::ccl_kernel(vecmem::data::vector_view<const traccc::cell>, vecmem::data::vector_view<const traccc::cell_module>, unsigned short, unsigned short, vecmem::data::vector_view<traccc::measurement>, unsigned int &, vecmem::data::vector_view<unsigned int>) (314, 1, 1)x(128, 1, 1), Device 0, CC 8.6, Invocations 110
    Section: Command line profiler metrics
    ---------------------- ----------- ---------- ---------- ----------
    Metric Name            Metric Unit    Minimum    Maximum    Average
    ---------------------- ----------- ---------- ---------- ----------
    gpu__time_duration.sum     nsecond 367,424.00 377,664.00 371,650.91
    ---------------------- ----------- ---------- ---------- ---------- 
```

After this PR (commit `35f10fa2`):

```
  traccc::cuda::kernels::ccl_kernel(vecmem::data::vector_view<const traccc::cell>, vecmem::data::vector_view<const traccc::cell_module>, unsigned short, unsigned short, vecmem::data::vector_view<traccc::measurement>, vecmem::data::vector_view<unsigned int>) (314, 1, 1)x(128, 1, 1), Device 0, CC 8.6, Invocations 110
    Section: Command line profiler metrics
    ---------------------- ----------- ------------ ------------ ------------
    Metric Name            Metric Unit      Minimum      Maximum      Average
    ---------------------- ----------- ------------ ------------ ------------
    gpu__time_duration.sum     nsecond 1,490,784.00 1,508,992.00 1,499,022.55
    ---------------------- ----------- ------------ ------------ ------------
```

I.e. a slowdown of about 4 times. :(

Reply:
Also if you want to rewind the memory of the measurement sorting, please take a look into the past PR:
https://github.com/acts-project/traccc/pull/467

Reply:
Thanks for being aware of that!

Reply:
I agree it's a bit mysterious... I think it might be worth investigating the contention on atomics in global memory. NSight Compute allows you to see why threads are stalling. If the number of long scoreboard stalls - which I think should include atomics - is higher in the updated version, that could explain it.

Reply:
Suspiciously, the version in this PR (blue) simply performs way more instructions than the version before this PR (green). :thinking: Are we doing some duplicate work?

![image](https://github.com/acts-project/traccc/assets/1165820/f7511d4e-4adc-4246-9d6f-c6a96370ea36)